p2pclient/unary_handlers.go

```diff
@@ -0,0 +1,279 @@
+
+}
+
+func (c *Client) getPersistentWriter() ggio.WriteCloser {
+	c.openPersistentConn.Do(
+		func() {
+			conn, err := c.newControlConn()
+			if err != nil {
+				panic(err)
+			}
+
+			w := NewSafeWriter(ggio.NewDelimitedWriter(conn))
+			w.WriteMsg(&pb.Request{Type: pb.Request_PERSISTENT_CONN_UPGRADE.Enum()})
+			c.persistentConnWriter = w
+
+			r := ggio.NewDelimitedReader(conn, network.MessageSizeMax)
+			go c.run(r, c.persistentConnWriter)
+
+		},
+	)
+
+	return c.persistentConnWriter
+}
+
+func (c *Client) getResponse(callID uuid.UUID) (*pb.PersistentConnectionResponse, error) {
+	rc, _ := c.callFutures.LoadOrStore(callID, make(persistentConnectionResponseFuture))
+	defer c.callFutures.Delete(callID)
+
+	response := <-rc.(persistentConnectionResponseFuture)
+	if dErr := response.GetDaemonError(); dErr != nil {
+		return nil, newDaemonError(dErr)
+	}
+
+	return response, nil
+}
+
+func (c *Client) AddUnaryHandler(proto protocol.ID, handler UnaryHandlerFunc) error {
+	w := c.getPersistentWriter()
+
+	callID := uuid.New()
+
+	w.WriteMsg(
+		&pb.PersistentConnectionRequest{
+			CallId: callID[:],
+			Message: &pb.PersistentConnectionRequest_AddUnaryHandler{
+				AddUnaryHandler: &pb.AddUnaryHandlerRequest{
+					Proto: (*string)(&proto),
+				},
+			},
+		},
+	)
+
+	if _, err := c.getResponse(callID); err != nil {
+		return err
+	}
+
+	c.mhandlers.Lock()
+	c.unaryHandlers[proto] = handler
+	c.mhandlers.Unlock()
+
+	return nil
+}
+
+func (c *Client) CallUnaryHandler(
+	ctx context.Context,
+	peerID peer.ID,
+	proto protocol.ID,
+	payload []byte,
+) ([]byte, error) {
+
+	w := c.getPersistentWriter()
+
+	callID := uuid.New()
+
+	// both methods don't return any errors
+	cid, _ := callID.MarshalBinary()
+	pid, _ := peerID.MarshalBinary()
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+
+		select {
+		case <-done:
+			return
+		case <-ctx.Done():
+			w.WriteMsg(
+				&pb.PersistentConnectionRequest{
+					CallId:  cid,
+					Message: &pb.PersistentConnectionRequest_Cancel{Cancel: &pb.Cancel{}},
+				},
+			)
+		}
+	}()
+
+	w.WriteMsg(
+		&pb.PersistentConnectionRequest{
+			CallId: cid,
+			Message: &pb.PersistentConnectionRequest_CallUnary{
+				CallUnary: &pb.CallUnaryRequest{
+					Peer:  pid,
+					Proto: (*string)(&proto),
+					Data:  payload,
+				},
+			},
+		},
+	)
+
+	response, err := c.getResponse(callID)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.GetCancel() != nil {
+		return nil, ctx.Err()
+	}
+
+	result := response.GetCallUnaryResponse()
+	if len(result.GetError()) != 0 {
+		return nil, newP2PHandlerError(result)
+	}
+
+	select {
+	case done <- struct{}{}:
+		return result.GetResponse(), nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (c *Client) Cancel(callID uuid.UUID) error {
+	return c.getPersistentWriter().WriteMsg(
+		&pb.PersistentConnectionRequest{
+			CallId: callID[:],
+			Message: &pb.PersistentConnectionRequest_Cancel{
+				Cancel: &pb.Cancel{},
+			},
+		},
+	)
+}
+
```

Comment:
```suggestion
```

Please either remove this function or call it inside `CallUnaryHandler()` (currently, there is duplicated code).

I prefer the first option since this function is unlikely to be called from the outside, since we don't expose call IDs to the user.